packages/wells/src/client/api/wellbores.ts

```diff
@@ -32,18 +35,56 @@ export class Wellbores extends AssetsAPI {
    */
   private mapToWellbore = async (assets: Asset[]): Promise<Wellbore[]> => {
     return assets.map(asset => {
-      return <Wellbore>{
+      const wellbore = <Wellbore>{
         id: asset.id,
         name: asset.name,
-        parentId: asset.parentId,
         metadata: asset.metadata,
         trajectories: async (): Promise<Survey[]> => {
           return await this.surveys.listTrajectories(asset.id);
         },
+        parentWell: async (): Promise<Well | undefined> => {
+          return await this.getParentWell(asset);
+        },
+      };
+      wellbore.getDatum = async (): Promise<WellDatum | undefined> => {
+        const well = await wellbore.parentWell();
+        if (well) {
+          return well.datum;
+        }
+      };
+      wellbore.getWellHeadLocation = async (): Promise<
+        WellHeadLocation | undefined
+      > => {
+        const well = await wellbore.parentWell();
+        if (well) {
+          return well.wellHeadLocation;
+        }
       };
+      return wellbore;
     });
   };
 
+  /**
+   * Method for finding the parent wellbore of a given wellbore asset
+   *
+   * ```
+   * const parent = await client.wellbores.getParentWell(asset);
+   * ```
+   * @param asset
+   */
+  private getParentWell = async (asset: Asset): Promise<Well | undefined> => {
+    if (asset.parentId) {
+      let parentWell = await this.retrieve([{ id: asset.id }]);
+      while (parentWell[0].metadata && parentWell[0].metadata.type !== 'Well') {
+        if (!asset.parentId) {
+          return undefined;
+        }
+        parentWell = await this.retrieve([{ id: asset.parentId }]);
+      }
+      return Wells.mapToWell(parentWell)[0];
```

Comment:
So, my understanding is that you search upwards in the asset hierarchy until you find a parent-asset that is type `'Well` and return that well. I guess a wellbore cannot keep a parent reference to another wellbore? so it has to have a reference to parent that is a type 'Well'. If there are any parent If the user insert an asset that has type 'well', you should just return that asset as a Well.

I guess that if the argument asset is already type 'Well', you could just check that on the first line and return `undefined `straight away as `if (asset.metadata.type == 'Well') return undefined `

To be able to hit the condition on line 79: `if (!asset.parentId)` the condition `if (asset.parentId)` would have to be true. So, I guess the condition on line 79 will always be false?

on line 77, if you want to get the `parentWell`, shouldn't it be: `let parentWell = await this.retrieve([{ id: asset.parentId }]);`? If not, you could just do `let parentWell = asset` 

On line 82, let's say something has gone wrong in the ingestion and the type of the parent asset (parentWell[0].metadata.type = 'Wel') or something else. I guess the programming will get stuck in the while-loop infinitely?

Reply:
A well can't have another well as a parent, so yeah I think if it has type well I can just return the well, and any other type than wellbore I will just return undefined.

If something has gone wrong in the ingestion then the loop will continue to run past the 'Wel' asset and get to the top of the asset heirarchy where the parentId will be undefined, and therefore the loop will return undefined.

The queries you had about line 76 and 79, and line 77 are typos, which I can fix

Reply:
with regards to the 'Wel' example, my understanding is that the results from `parentWell = await this.retrieve([{ id: asset.parentId }]);` will never change, since the id in `asset.parentId` will always be the same. It would run past 'Wel' if it was `parentWell = await this.retrieve([{ id: parentWell.parentId }]);`?

Reply:
Oh that's another typo! I guess it's not caught in the tests because the wellbore tested on is a direct descendant of the well

Reply:
Nice catch!